### DISCUSSION TITLE
Possible improvements and ideas

### DISCUSSION THREAD

**Discussion by alextrical** [aamott/petinator#14] [General]
I have a few ideas from a similar project that could be of use to you in this project.

For the menu, take a look at TCMenu, it can allow you to design the menu in a GUI and export it to a sketch.
Maybe try using a OLED display such as the SSD1306, its small at 1.3" but gives a lot of resolution, and is pretty cheap

Good find with the [lbussy/QuickPID](https://github.com/lbussy/QuickPID) by the way, I will give that a try shortly.
After that I need to look into getting Thermal runaway detection bugs ironed out, and also how to get it to compile with a CICD pipeline in GitHub

**Comment by alextrical** [comment]
I have been working on custom component in the Wokwi system, and it seems that it is possible to create a part that approximates the feedback loop of a hot end. Converting PWM signals into a calculated energy, and using that to calculate a change in temperature for a given mass of aluminium. Adding in some loss to approximate a real world cenario, it's working rather well.

The issue I'm hitting now is using an interpolated lookup table to convert the values back to what is expected for the thermistor used in the released code. Specifically a 104GT-2

**Reply by alextrical** [reply]
This is the results of my experiments so far.
With this, we now have something that models the 104GT-2 resistance curve (using Steinhart-Hart coefficients for both reading and simulating), and approximates a thermal mass of a hot end (fully adjustable Metal material, Mass, Wattage, thermal losses, Thermistor coefficients, room temperature)
https://wokwi.com/projects/356586429350093825

**Comment by aamott** [comment]
TCMenu was actually one of the first menus I looked at and strongly considered. However, until interrupts can be implemented for the stepper on a large number of boards (see #11), it's got to it work in one loop, which means a very optimized menu. I wrote an optimized menu library in the [timer](https://github.com/aamott/petinator/tree/timer-free) branch, but an upgrade might be good once a good timer interrupt solution can be figured out. 

I had some issues with thermal runaway detection as well. I'll push up a few of my changes soon. 

I'd heard of custom components in Wokwi, but never played with them. That's some awesome work! Much better than the wrong thermistor and an LED.

**Reply by alextrical** [reply]
The cross platform ISR library you linked looks really nice, I will take a look into that.
It seems we are both aiming for the same goal of cross platform support, but starting in different places. I went with the cheapest off the shelf board i could find (Specifically the $15 BTT EBB42) to try and lower the cost of entry for general users not wanting to build from component level parts. Though the idea is expand platform support once the design matured.
It also had the up shot that the driver was a TMC2209 and didn't need a PWM signal to be driven, as the driver has an internal oscillator that can be set over UART.

If you would be interested in trying the board, we can look into getting one sent over.

I'm hoping to get some progress adapting/isolating the PID and thermal runaway logic from Marlin into a stand alone example, in the next couple of weeks

**Comment by aamott** [comment]
Looks like the new stepper library is stable on #17. I think adding support for the TMC2209 is a great next step! Is the BTT EBB42 an STM32 with a TMC2209? That could be a good board to make this easier. Any changes should probably be made to `stepper.h` (and really, the changes I made using `FastAccelStepper` should probably have gone there too...) but if you already have the board, that would be great to experiment with. I may end up getting one soon.
Here's a great library to start with: [TMC2209](https://github.com/janelia-arduino/TMC2209)

**Reply by alextrical** [reply]
Sorry about the delayed response, seems that I missed this response (only just figured out GitHub has a home page to see all Discussions 🤦‍♂️)

Yes the EBB42/EBB36 are toolhead boards designed for Klipper, containg the TMC2209 an STM32G0B1CBT, 2 fan controllers, a PTC socket, a RTD socket (on the upgraded board) and a MOSFET socket for a heater

There is minimal IO exposed, but there are 8 GPIO (one of which is hidden behind a Schmitt trigger) and 4 of them have pull up resistors.

The display I have been working on (Dinky Display) is a I2C SSD1306, port expander for a rotary encoder and EEPROM to round off the missing features while still keeping the total cost low and accessable![20230521_174022.jpg](https://github.com/aamott/petinator/assets/35117191/1c2c0621-e9f3-439a-9e48-99f1b4703972)

With the FW I've been working on, I've got it auto compiling via GitHub CICD, working with PlatformIO and platform agnostic.
Auto detecting the presence of the I2C expander and changing the input source based on that, or fall back to 'endstop' pins.
Im also working on merging in a system to identify the attached I2C display and change to the correct U8G2 definition (for SSD1306, SSD1309, SH1106 and SH1107), just having issues working out how to make that work with TCMenu

 The big issue I have currently is a good PID tune system

**Reply by alextrical** [reply]
As for the stepper library for the TMC family of steppers, the main one to use is [TMCStepper](https://github.com/teemuatlut/TMCStepper)

It's a very mature library as it's the one used by Marlin and had more surface to be bugfixed.
Control over the motor can be entirely manipulated over the Half Duplex UART line, or you can control it via toggling pins/PWM

**Reply by aamott** [reply]
Very nice work! I've gotten a little busy recently and haven't had time for this project, but I'm planning to get back to work soon. I started adding support for TMC drivers through that same library, but I still haven't pushed the branch up. I'll try to do that soon and get back on this.
